File: packages/core/src/scheduler.ts
import { db } from "./db";
import { getEntity, getVerb } from "./repo";
import { evaluate, createScriptContext } from "@viwo/scripting";

/**
 * Manages scheduled tasks (delayed verb executions).
 * Tasks are persisted in the database until execution.
 */
export class TaskScheduler {
  constructor() {}

  /**
   * Schedules a verb to be executed on an entity after a delay.
   *
   * @param entityId - The ID of the entity to execute the verb on.
   * @param verb - The name of the verb.
   * @param args - Arguments to pass to the verb.
   * @param delayMs - Delay in milliseconds.
   */
  schedule(
    entityId: number,
    verb: string,
    args: readonly unknown[],
    delayMs: number,
  ) {
    const executeAt = Date.now() + delayMs;
    db.query(
      "INSERT INTO scheduled_tasks (entity_id, verb, args, execute_at) VALUES (?, ?, ?, ?)",
    ).run(entityId, verb, JSON.stringify(args), executeAt);
  }

  private sendFactory: (
    entityId: number,
  ) => (type: string, payload: unknown) => void = () => () => {};
  /**
   * Sets the factory function for creating the 'send' function used in scheduled tasks.
   * This allows the scheduler to send messages to clients even when triggered asynchronously.
   *
   * @param factory - A function that returns a send function for a given entity ID.
   */
  setSendFactory(
    factory: (entityId: number) => (type: string, payload: unknown) => void,
  ) {
    this.sendFactory = factory;
  }

  private intervalId: Timer | null = null;

  /**
   * Starts the scheduler loop.
   *
   * @param intervalMs - The interval in milliseconds to check for due tasks.
   */
  start(intervalMs: number = 1000) {
    if (this.intervalId) return;
    this.intervalId = setInterval(() => this.process(), intervalMs);
    console.log(`[Scheduler] Started with interval ${intervalMs}ms`);
  }

  /**
   * Stops the scheduler loop.
   */
  stop() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
      console.log("[Scheduler] Stopped");
    }
  }

  /**
   * Processes all due tasks.
   * Should be called periodically (e.g., every tick).
   */
  async process() {
    const now = Date.now();
    const tasks = db
      .query("SELECT * FROM scheduled_tasks WHERE execute_at <= ?")
      .all(now) as any[];

    if (tasks.length === 0) return;

    // Delete tasks immediately
    const ids = tasks.map((t) => t.id);
    db.query(
      `DELETE FROM scheduled_tasks WHERE id IN (${ids.join(",")})`,
    ).run();

    if (!this.sendFactory) {
      throw new Error("[Scheduler] No send factory set.");
    }

    for (const task of tasks) {
      // Create a send function specific to this entity

      try {
        const entity = getEntity(task.entity_id);
        const verb = getVerb(task.entity_id, task.verb);
        const args = JSON.parse(task.args);

        const send = this.sendFactory(task.entity_id);

        if (entity && verb) {
          await evaluate(
            verb.code,
            createScriptContext({
              caller: entity, // System is caller? Or self?
              this: entity,
              args,
              send,
            }),
          );
        }
      } catch (e) {
        console.error(`[Scheduler] Error executing task ${task.id}:`, e);
      }
    }
  }
}

export const scheduler = new TaskScheduler();

File: crates/viwo-core/src/scheduler.rs
//! Task scheduler for delayed verb execution.
//!
//! The scheduler manages tasks stored in the database and executes them
//! when their scheduled time arrives. Tasks are persisted to survive restarts.

use crate::{StorageError, WorldStorage};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use thiserror::Error;
use tokio::sync::Mutex;
use tokio::time;

#[derive(Debug, Error)]
pub enum SchedulerError {
    #[error("Storage error: {0}")]
    Storage(#[from] StorageError),

    #[error("Task execution error: {0}")]
    Execution(String),
}

// Re-export ScheduledTask from storage for convenience
pub use crate::storage::ScheduledTask;

/// Task scheduler that executes verbs after a delay.
pub struct Scheduler {
    storage: Arc<Mutex<WorldStorage>>,
    interval_ms: u64,
}

impl Scheduler {
    /// Create a new scheduler.
    ///
    /// # Arguments
    /// * `storage` - Shared world storage
    /// * `interval_ms` - How often to check for due tasks (in milliseconds)
    pub fn new(storage: Arc<Mutex<WorldStorage>>, interval_ms: u64) -> Self {
        Self {
            storage,
            interval_ms,
        }
    }

    /// Schedule a task to execute after a delay.
    ///
    /// # Arguments
    /// * `entity_id` - Entity to execute the verb on
    /// * `verb` - Name of the verb to execute
    /// * `args` - Arguments to pass to the verb (as JSON Value)
    /// * `delay_ms` - Delay in milliseconds before execution
    pub async fn schedule(
        &self,
        entity_id: i64,
        verb: &str,
        args: serde_json::Value,
        delay_ms: u64,
    ) -> Result<i64, SchedulerError> {
        let execute_at = (current_time_ms() + delay_ms) as i64;
        let storage = self.storage.lock().await;
        let task_id = storage.schedule_task(entity_id, verb, args, execute_at)?;
        Ok(task_id)
    }

    /// Get all tasks that are due for execution.
    async fn get_due_tasks(&self) -> Result<Vec<ScheduledTask>, SchedulerError> {
        let now = current_time_ms() as i64;
        let storage = self.storage.lock().await;
        let tasks = storage.get_due_tasks(now)?;
        Ok(tasks)
    }

    /// Delete a task from the database.
    async fn delete_task(&self, task_id: i64) -> Result<(), SchedulerError> {
        let storage = self.storage.lock().await;
        storage.delete_task(task_id)?;
        Ok(())
    }

    /// Process all due tasks.
    ///
    /// This should be called periodically by the server. Tasks are executed
    /// by calling the provided execution callback.
    pub async fn process<F, Fut>(&self, mut execute: F) -> Result<(), SchedulerError>
    where
        F: FnMut(ScheduledTask) -> Fut,
        Fut: std::future::Future<Output = Result<(), String>>,
    {
        let tasks = self.get_due_tasks().await?;
        if tasks.is_empty() {
            return Ok(());
        }

        // Execute and delete tasks one by one
        for task in tasks {
            // Delete task before executing to avoid re-execution on failure
            self.delete_task(task.id).await?;

            if let Err(e) = execute(task.clone()).await {
                eprintln!(
                    "[Scheduler] Error executing task {} (entity {}, verb {}): {}",
                    task.id, task.entity_id, task.verb, e
                );
            }
        }

        Ok(())
    }

    /// Run the scheduler loop.
    ///
    /// This continuously checks for due tasks at the configured interval
    /// and executes them using the provided callback.
    pub async fn run<F, Fut>(self: Arc<Self>, execute: F)
    where
        F: Fn(ScheduledTask) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<(), String>> + Send,
    {
        let mut interval = time::interval(Duration::from_millis(self.interval_ms));
        let execute = Arc::new(execute);

        loop {
            interval.tick().await;

            let exec_clone = Arc::clone(&execute);
            if let Err(e) = self.process(|task| exec_clone(task)).await {
                eprintln!("[Scheduler] Error processing tasks: {}", e);
            }
        }
    }
}

/// Get current time in milliseconds since Unix epoch.
fn current_time_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("System time before Unix epoch")
        .as_millis() as u64
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_schedule_and_retrieve() {
        let storage = Arc::new(Mutex::new(WorldStorage::in_memory().unwrap()));
        let scheduler = Scheduler::new(Arc::clone(&storage), 100);

        // Create an entity
        let entity_id = {
            let storage = storage.lock().await;
            storage
                .create_entity(serde_json::json!({"name": "Test"}), None)
                .unwrap()
        };

        // Schedule a task for immediate execution
        scheduler
            .schedule(entity_id, "test_verb", serde_json::json!(["arg1", 42]), 0)
            .await
            .unwrap();

        // Wait a tiny bit to ensure time passes
        tokio::time::sleep(Duration::from_millis(10)).await;

        // Get due tasks
        let tasks = scheduler.get_due_tasks().await.unwrap();
        assert_eq!(tasks.len(), 1);
        assert_eq!(tasks[0].entity_id, entity_id);
        assert_eq!(tasks[0].verb, "test_verb");
        assert_eq!(tasks[0].args.as_array().unwrap().len(), 2);
    }

    #[tokio::test]
    async fn test_process_executes_and_deletes() {
        let storage = Arc::new(Mutex::new(WorldStorage::in_memory().unwrap()));
        let scheduler = Scheduler::new(Arc::clone(&storage), 100);

        let entity_id = {
            let storage = storage.lock().await;
            storage
                .create_entity(serde_json::json!({"name": "Test"}), None)
                .unwrap()
        };

        // Schedule a task
        scheduler
            .schedule(entity_id, "greet", serde_json::json!([]), 0)
            .await
            .unwrap();

        tokio::time::sleep(Duration::from_millis(10)).await;

        // Process tasks with a simple callback
        let mut executed = false;
        scheduler
            .process(|task| {
                executed = true;
                assert_eq!(task.verb, "greet");
                async { Ok(()) }
            })
            .await
            .unwrap();

        assert!(executed, "Task should have been executed");

        // Verify task was deleted
        let tasks = scheduler.get_due_tasks().await.unwrap();
        assert_eq!(tasks.len(), 0, "Task should be deleted after execution");
    }

    #[tokio::test]
    async fn test_only_executes_due_tasks() {
        let storage = Arc::new(Mutex::new(WorldStorage::in_memory().unwrap()));
        let scheduler = Scheduler::new(Arc::clone(&storage), 100);

        let entity_id = {
            let storage = storage.lock().await;
            storage
                .create_entity(serde_json::json!({"name": "Test"}), None)
                .unwrap()
        };

        // Schedule task far in the future
        scheduler
            .schedule(entity_id, "future_task", serde_json::json!([]), 10_000)
            .await
            .unwrap();

        // Process - should not execute anything
        let mut executed = false;
        scheduler
            .process(|_task| {
                executed = true;
                async { Ok(()) }
            })
            .await
            .unwrap();

        assert!(!executed, "Future task should not execute yet");

        // Task should still be in database
        let tasks = scheduler.get_due_tasks().await.unwrap();
        assert_eq!(tasks.len(), 0, "Future task not yet due");
    }
}
